### DISCUSSION TITLE
Sniprun AUR package

### DISCUSSION THREAD

**Discussion by lmartinez-mirror** [michaelb/sniprun#58] [General]
Hi, I'm currently writing up several PKGBUILDs of sniprun for submitting into the AUR. One of the things Arch Linux has in their guidelines for Rust packages is [using a Cargo.lock file for compiling.](https://wiki.archlinux.org/index.php/Rust_package_guidelines#Building)

It would be great to have an up-to-date Cargo.lock file in the current and future releases for this.

**Comment by michaelb** [comment]
Sure, i guess some dependencies are not quite up to date in there.

I'll try to keep them up to date manually, since i guess that "*" dependencies are so-so and that there is not way to automate that like the pkgver.....but if there is a way, i'll take it :-). 

Maybe a github bot? Or maybe "*" dependencies in cargo.toml are ok, and then updating the lock is only one 'cargo update' away? You tell me


I also use Arch and i have a few packages up there myself


I think what need some attention is the (new) issue with neovim 0.4.x vs 0.5 stuff.
The install.sh script is basically patchwork, with different failsafes etc... but the important compatibility stuff lies in "neovim < 0.5 needs sniprun < 0.5"

**Reply by michaelb** [reply]
God i'm so ashamed of this script

**Reply by lmartinez-mirror** [reply]
If it means anything, I don't use the install.sh for the PKGBUILD, opting to place things manually with `install`. Regarding the compatibility issues, I'd have to create a separate PKGBUILD for those using `neovim` stable. It shouldn't be too much of a hassle.

As for the Cargo.lock, do what is easiest for you. I don't have the slightest clue about Rust and their file trees.

**Reply by michaelb** [reply]
Since i won't be publishing a crate, I went with the easy solution: "*" dependencies (meaning i'll fetch latest at each `cargo update` which will then put fixed version in the Cargo.lock).

The Cargo.lock file is now checked-in with git and up to date, i'll keep cargo-updating at each release.

**Comment by lmartinez-mirror** [comment]
This is what I have thus far. It's a split package because the vim and Rust components come from the same upstream.

```sh
# Maintainer: lmartinez-mirror
pkgname=('sniprun' 'neovim-sniprun')
pkgver=0.5.2
pkgrel=1
pkgdesc='A neovim plugin to independently run snippets of code'
arch=('x86_64')
url="https://github.com/michaelb/sniprun"
license=('MIT')
makedepends=('cargo' 'gcc-libs')
source=("$pkgname-$pkgver.tar.gz::$url/archive/refs/tags/v$pkgver.tar.gz")
sha256sums=('97d952f54235cf5b52553cafbb9466b2f83dbc1c562279b9ce940dfc308a00c7')

build() {
  cd "$pkgname-$pkgver"
  cargo build --release --all-features --target-dir=target
}

# check() {
#   cd "$pkgname-$pkgver"
#   cargo test --release --locked --target-dir=target
# }

package_sniprun() {
  pkgdesc='Compiled binary core for neovim-sniprun'
  depends=('gcc-libs')

  cd "$pkgname-$pkgver"
  install -Dm 755 target/release/sniprun -t "$pkgdir/usr/bin/"
  install -Dm 644 LICENSE -t "$pkgdir/usr/share/licenses/$pkgname/"
  install -Dm 644 README.md -t "$pkgdir/usr/share/doc/$pkgname/"
}

package_neovim-sniprun()  {
  arch=('any')
  depends=('neovim-git' 'sniprun')
  optdepends=('bash: Bash snippet support'
              'gcc: C snippets support'
              'gcc-ada: Limited Ada snippet support'
              'ghc: Haskell snippets support'
              'g++: C++ snippets support'
              'python3: Python3 snippets support'
              'jupyter: Jupyter support'
              'jupyter_console: Required with jupyter')
  install="$pkgname.install"

  cd "sniprun-$pkgver"
  find autoload doc plugin lua \
    -type f -exec install -Dm 644 '{}' "$pkgdir/usr/share/nvim/runtime/{}" \;
  install -d "$pkgdir/usr/share/nvim/runtime/target/release"
  ln -s "/usr/bin/sniprun" "$pkgdir/usr/share/nvim/runtime/target/release/sniprun"
  install -d "$pkgdir/usr/share/licenses/"
  install -d "$pkgdir/usr/share/doc/"
  ln -s "/usr/share/licenses/sniprun/" "$pkgdir/usr/share/licenses/$pkgname"
  ln -s "/usr/share/doc/sniprun" "$pkgdir/usr/share/doc/$pkgname"
}
```

**Reply by michaelb** [reply]
looks good!

Frankly, I'm just a little bit concerned about the line that puts autoload, doc, plugin and lua in neovim runtime (through a link i guess?). If that's how other AUR-packaged neovim plugin do, it should be alright though.
You may also be missing a `--locked` in the build section

Also, if I were you, i'd forget about all those optdepends, (with the exception of gcc)  because they are subject to change, additions, and even incompatible version requirements depending on which interpreter is selected within sniprun. You already miss a lot of them (java, scala, julia, R, etc...) and it does not make much sense: if anything needs attention, the user can see it in doc/




You will probably be asked to add a pgkver() function at some point that should look like this. This will automate updating the version on the package along with the latest git tag, and will be run a makepkg time.

```
pkgver() {
  cd "${pkgname}"
  git describe --long --tags | sed 's/^v//;s/\([^-]*-g\)/r\1/;s/-/./g'
}
```

**Reply by lmartinez-mirror** [reply]
Thanks for the tips!

The `find` command for autoload, doc, etc. is what I use to properly install the directories and their contents for the package.

I think I'll keep the optdepends for now. I need to make sure the docs are readable when installed by package; I haven't been able to get a (neo)vim package that can read the documentation with `:help` for whatever reason. Besides, if there's a package that complements `sniprun`, it would be best to add it to optdepends; if an interpreter *does* get removed, I'll remove it from there at next release.

I have another PKGBUILD (`sniprun-git`) that does have a `pkgver()`. This PKGBUILD doesn't have one because it targets stable releases only. It's this one where I won't add the extra optdepends due to future changes with interpreters.

**Reply by michaelb** [reply]
oh rigth.

Also, neovim <0.5 users are missing on that feature, but post 0.5.1 docs can be consulted from `:SnipInfo <tab-completed name>`, and the access is done via relative path, so it's important the directories in the project root are kept 'next' to each other.

Those were initially made to be consulted online though:
I don't think the files in there, apart from sniprun.txt, are suitable in content or form to be accessed from `:help`, hence their .md format (and not .txt).

**Reply by lmartinez-mirror** [reply]
One last thing!

To-date, I have at least four (soon to be five) PKGBUILDs for the following:

* `sniprun` and `neovim-sniprun`
* `sniprun-bin` and `neovim-sniprun-bin` for binary-only releases (the neovim package is identical to `neovim-sniprun`
* `sniprun-git` and `neovim-sniprun-git` for tracking master branch
* `sniprun-legacy` and `neovim-sniprun-legacy` targeting version 0.4.9 for neovim < 0.5.0
* And soon `sniprun-legacy-bin` and `neovim-sniprun-legacy-bin` for binary-only version 0.4.9 for neovim < 0.5.0

Can the sniprun plugin run with an older core with the caveat of missing features, or are they tied to their respective release? Example: Plugin version 0.5.2 loads a sniprun core < 0.5.0

**Reply by michaelb** [reply]
Okay, that's a lot of packages 😅. Just being curious, what's the difference between sniprun and neovim-sniprun? One's the core and the other is the plugin stuff ?

First, mind that the master and releases have always been close to each other and will continue to be so.
I mean that they are in very much sync because most of the developpent in done on dev.
Maybe that lessen the importance of a 'git' release pointing to master




Sniprun 0.5 introduced breaking changes in both sides of the communication nvim <-> core (to simplify). So no, you can't load an old core with the new plugin or vice versa.

What the installer does, is , basically: checks for the neovim version, if 0.4.x,  'git reset' to the sniprun v0.4.9 tag, effectively downgrading the 'plugin' part and fetching the 0.4.9 binary.
That way it works ok when neovim 0.4.x users clone and ./install.sh the new releases, because they will just downgrade to 0.4.9

**Reply by lmartinez-mirror** [reply]
Sorry for the late reply.

Yes, `sniprun` and `neovim-sniprun` are the core and plugin, respectively.

If master is more of a "stable" branch, then I could track dev for `sniprun-git` instead. I know there will be *some* users who will want to be on the bleeding edge.

I have gone and added the missing optdepends to the stable PKGBUILD, but omitted them from the `-git` releases. I'll start uploading `sniprun`  shortly, with the rest to follow once I have them sorted out.

EDIT: Now I remember why I haven't uploaded yet. The docs aren't readable through neovim due to lack of helptags. The package can't generate them either because the neovim packages own the file, and that would cause a file conflict.

**Reply by michaelb** [reply]
tracking dev for `-git` seems more legit, i'm on dev myself.


What doesn't work exactly? 

`:help sniprun`

or 

`:SnipInfo <interpreter_name>`


If that's the latter, I may be able to do some tricks, like: if not found in the relative path, look in some hardcoded absolute path location


(And don't worry about replying late, we probably aren't in similar timezones, it's perfectly understandable)

**Reply by lmartinez-mirror** [reply]
`:help sniprun` doesn't work, it's an Arch Linux issue. It's nothing either of us can really fix, so I wouldn't worry about it.

**Reply by michaelb** [reply]
actually, if `:SnipInfo <interpreter_name>` works, it can be made to display the general help as well

**Comment by michaelb** [comment]
I've tested the AUR version of sniprun, it works great.
However, some bad decisions of mine (**now "fixed"**, in dev and will be in v0.5.3) about error handling mean that the absence of the ressources/ directory path relative to the core will crash `:SnipInfo`.

 It's not a problem for general usage though, sniprun will restart by itself at the next call.



The message about the bash interpreter is very thoughful.

**Comment by lmartinez-mirror** [comment]
This is a notice that I will be removing `sniprun-legacy` from the AUR after the main `neovim` package in the community repos updated to 0.5.0 a few weeks ago. I will also be listing it under the main `sniprun`'s `replaces` array for those who might be using it.

**Reply by michaelb** [reply]
Sure! The 'legacy' sniprun does not make sense anymore for Arch users.


Thanks for your involvement in this !
